README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# entrainment

<!-- badges: start -->

[![Project Status: WIP – Initial development is in progress, but there
has not yet been a stable, usable release suitable for the
public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
[![Documentation
Status](https://readthedocs.org/projects/entrainment/badge/?version=latest)](https://entrainment.readthedocs.io/en/latest/?badge=latest)
[![License:
MIT](https://img.shields.io/badge/license-MIT-green)](https://choosealicense.com/licenses/mit/)
[![Contributor
Covenant](https://img.shields.io/badge/Contributor%20Covenant-v2.0%20adopted-ff69b4.svg)](https://www.contributor-covenant.org/version/2/1/code_of_conduct/)
<!-- badges: end -->

<!-- TODO: Add Code Coverage, Build check, fair-software.eu, and CII Best Practices -->

## Overview

`entrainment` is a rule-based model created on Python to test and to
demonstrate the 24h light/dark cycle [entrainment
phenomenon](https://en.wikipedia.org/wiki/Entrainment_(chronobiology)).

## Prerequisites

You need to have some familiarity with the [Python programming
language](https://www.python.org/) to use `entrainment` main functions.

In case you don’t feel comfortable with Python, we strongly recommend
checking Jake VanderPlas free and online book [Python Data Science
Handbook](https://jakevdp.github.io/PythonDataScienceHandbook/) and the
Coursera course from Google [Crash Course on
Python](https://www.coursera.org/learn/python-crash-course) (free for
audit students).

## Installation

You can install `entrainment` from GitHub with:

``` bash
pip install git+https://github.com/danielvartan/entrainment.git#egg=entrainment
```

We don’t intend to publish this package to [PyPI](https://pypi.org/).

## Usage

`entrainment` comes with 5 subpackages:

- `model`: The entrainment model.
- `labren`: Tools for accessing [LABREN’s global horizontal solar
  irradiation model](http://labren.ccst.inpe.br/atlas_2017.html).
- `hypothesis`: Tools for hypothesis testing.
- `demo`: Tools to plot the entrainment dynamics.
- `utils`: Utility tools.

You can learn more about each subpackage and its functions on the
[package documentation website](https://entrainment.readthedocs.io).

The following example illustrates how to run the model.

Please note that in this example all of the model arguments are
assigned. You don’t need to assign values to all arguments, you can just
use the default values. Check `run_model()` documentation to learn more.

``` python
import entrainment.model as model

turtles = model.run_model(
    n = 10**3, # Number of turtles/subjects to create
    tau_range = (23.5, 24.6), # Limits for assigning Tau values
    tau_mean = 24.15, # Mean value for the Tau distribution
    tau_sd = 0.2, # Standard deviation value for the Tau distribution
    k_range = (0.001, 0.01), # Limits for assigning the k values
    k_mean = 0.001, # Mean value for the k distribution
    k_sd = 0.005, # Standard deviation value for the k distribution
    lam_c = 3750, # Critical lambda value
    labren_id = 1, # LABREN id of the global horizontal irradiation means
    by = "season", # Series resolution (choices: "month", "season", "year")
    n_cycles = 3, # Number of cycles to run
    start_at = 0, # Index number indicating the start of the series
    repetitions = 10**2, # Number of repetitions
    plot = True # Boolean value indicating if the function must plot the results
    )
```

<img src="docs/source/_static/readme_usage-example-1.png" alt="24h light/dark cycle entrainment of a population located at the south of Brazil by season" width="100%" />

## Citation

If you use `entrainment` in your research, please consider citing it. We
put a lot of work to build and maintain a free and open-source Python
package. You can find the `entrainment` citation below.

    Vartanian, D. (2023). {entrainment}: A rule-based model of the 24h light/dark cycle entrainment phenomenon. (v. 0.0.0.9000). https://github.com/danielvartan/entrainment

A BibTeX entry for LaTeX users is

    @Unpublished{,
        title = {{entrainment}: A rule-based model of the 24h light/dark cycle entrainment phenomenon},
        author = {Daniel Vartanian},
        year = {2023},
        url = {https://github.com/danielvartan/entrainment},
        note = {(v. 0.0.0.9000). Lifecycle: experimental},
    }

## Contributing

We welcome contributions, including bug reports.

Take a moment to review our [Guidelines for
Contributing](https://entrainment.readthedocs.io/en/latest/contributing.html).

## Acknowledgments

The initial development of `entrainment` was supported by a scholarship
provided by the [University of Sao Paulo (USP)](http://usp.br/) (❤️).

This model was initially created for the [SCX5002 - Complex System
I](https://uspdigital.usp.br/janus/Disciplina?tipo=D&sgldis=SCX5002&nomdis=&origem=C)
class of the [Graduate Program in Modeling Complex Systems
(PPG-SCX)](https://www.prpg.usp.br/pt-br/faca-pos-na-usp/programas-de-pos-graduacao/621-modelagem-de-sistemas-complexos)
of the [University of Sao Paulo (USP)](https://www5.usp.br/), under the
guidance of [Prof. Dr. Camilo Rodrigues
Neto](https://orcid.org/0000-0001-6783-6695).


CHANGELOG.md
<!--- https://py-pkgs.org -->

# entrainment (development version)

* Added a `CHANGELOG.md` file to track changes to the package.


py/model.py
## Source this file before using it.

import functools
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

def f_exact(lam, lam_c, k, tau, tau_ref = 24):
    logi_f = (tau_ref - tau) / (1 + np.exp(1) ** (- k * (lam - lam_c)))
    out = tau + logi_f

    return(out)

def exact(f, lam_c, k, tau, lam_0 = 0, lam_n = 10, h = 10 ** (- 3)):
    lam_list, y_list = [lam_0], [f(lam_0, lam_c, k, tau)]
    
    while lam_0 < lam_n:
        x_next = lam_0 + h
        y_next = f(x_next, lam_c, k, tau)
        lam_0 = x_next
        lam_list.append(x_next)
        y_list.append(y_next)
    
    return [lam_list, y_list]

def plot_exact(f, lam_c, k, tau, lam_0 = 0, lam_n = 10, 
               h = 10 ** (- 3)):
    data = exact(f, lam_c, k, tau, lam_0, lam_n)

    title = ("$\\lambda_c = {lam_c}$, $k = {k}$, $\\tau = {tau}$")\
         .format(lam_c = str(lam_c), k = str(k), tau = str(tau))

    plt.rcParams.update({'font.size': 10})
    plt.clf()
    
    fig, ax = plt.subplots()
    ax.plot(data[0], data[1], "r-", linewidth = 1)
    ax.set_xlabel("$\\lambda$")
    ax.set_ylabel("$f(\\lambda, \\lambda_{c}, k, \\tau)$")
    ax.set_title(title, fontsize = 10)
    plt.show()

plot_exact(f_exact, lam_c = 5, k = 2, tau = 22, lam_0 = 0, lam_n = 10)
plot_exact(f_exact, lam_c = 5, k = 2, tau = 26, lam_0 = 0, lam_n = 10)

def labren(id, name = None, by = "month",
           file = "./data/labren/global_horizontal_means.csv"):
    data = (
        pd.read_csv(filepath_or_buffer  = file, sep = ";")
        .rename(str.lower, axis = "columns")
        .loc[[id - 1]]
    )
    
    out = {"name": name}
    for i in list(data): out[i] = data.loc[id - 1, i]
    
    data = (
        data.iloc[:, 5:17]
        .transpose()
    )

    data.reset_index(inplace = True)
    data.columns = ["month", "x"]
    
    out["ts"] = list(data["x"])
    
    if by == "season":
        seasons = [
            ["Dec", "Jan", "Feb"], ["Mar", "Apr", "May"], ["Jun", "Jul", "Aug"], 
            ["Sep", "Oct", "Nov"]
            ]
        
        keys = ["name", "id", "country", "lon", "lat", "annual"]
        labels = ["Summer", "Autumn", "Winter", "Spring"]
        data = out
        out = {}
    
        for i, j in enumerate(data):
            if j in keys: out[j] = data[j]
        
        for i in range(4):
            mean = np.mean([data[i.lower()] for i in seasons[i]])
            out[labels[i].lower()] = mean
        
        out["ts"] = [out[i.lower()] for i in labels]
    
    return(out)

def plot_labren(id_1_index = 72272, id_2_index = 1,
                label_1 = "Nascente do rio Ailã (Lat.: $5.272$)",
                label_2 = "Arroio Chuí (Lat.: $- 33.752$)",
                by = "month"):
    if (by == "month"):
        x = [
            "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep",
            "Oct", "Nov", "Dec"
            ]
        
        y_1 = labren(id_1_index)["ts"]
        y_2 = labren(id_2_index)["ts"]
        xlabel = "Month"
    else:
        x = ["Summer", "Autumn", "Winter", "Spring"]
        y_1 = labren(id_1_index, by = "season")["ts"]
        y_2 = labren(id_2_index, by = "season")["ts"]
        xlabel = "Season"
    
    title = "Global Horizontal Solar Irradiation (Source: LABREN/INPE, 2017)"
    
    plt.rcParams.update({'font.size': 10})
    plt.clf()
    
    fig, ax = plt.subplots()
    ax.plot(x, y_1, "r-", label = label_1, linewidth = 1)
    ax.plot(x, y_2, "b-", label = label_2,  linewidth = 1)
    ax.set_xlabel(xlabel)
    ax.set_ylabel("$Wh / m^{2}.day$")
    ax.set_title(title, fontsize = 10)
    
    plt.legend(fontsize = 8)
    plt.show()

plot_labren(
    id_1_index = 72272, id_2_index = 1,
    label_1 = "Nascente do rio Ailã (Lat.: $5.272$)",
    label_2 = "Arroio Chuí (Lat.: $- 33.752$)",
    by = "month"
    )

plot_labren(
    id_1_index = 72272, id_2_index = 1,
    label_1 = "Nascente do rio Ailã (Lat.: $5.272$)",
    label_2 = "Arroio Chuí (Lat.: $- 33.752$)",
    by = "season"
    )

def entrain(lam, lam_c, k, tau, tau_ref = 24):
    logi_f = (tau_ref - tau) / (1 + np.exp(1) ** (- k * (lam - lam_c)))
    out = tau + logi_f
    error = np.random.uniform(low = 0, high = 1) * np.abs(out - tau)
    
    if out >= tau:
        out = out - error
    else:
        out = out + error
    
    return(out)

def entrain_turtles(turtles, turtles_0, lam, lam_c, lam_c_tol):
    n = len(turtles)
    out = []
    
    for i in range(n):
        tau_0 = turtles_0[i]["tau"]
        tau = turtles[i]["tau"]
        k = turtles[i]["k"]
        
        if (lam >= (lam_c - lam_c_tol)):
            tau_i = entrain(lam, lam_c, k, tau, tau_ref = 24)
        else:
            tau_i = entrain(lam, lam_c, k, tau, tau_ref = tau_0)
        
        out.append({"tau": tau_i, "k": k})
    
    return(out)

def create_turtles(n = 10 **2, tau_range = (23.5, 24.6), tau_mean = 24.15, 
                   tau_dp = 0.2, k_range = (0.001, 0.01), k_mean = 0.001, 
                   k_dp = 0.005):
    turtles = []
    
    for i in range(n):
        tau = np.random.normal(tau_mean, tau_dp)
        k = np.random.normal(k_mean, k_dp)
        
        if (tau < tau_range[0]): tau = tau_range[0]
        if (tau > tau_range[1]): tau = tau_range[1]
        if (k < k_range[0]): k = k_range[0]
        if (k > k_range[1]): k = k_range[1]
        
        turtles.append({"tau": tau, "k": k})
    
    return(turtles)

# np.array(labren(72272)["ts"]).mean() ~ 4727.833
def run_model(
    n = 10 ** 3, tau_range = (23.5, 24.6), tau_mean = 24.15, tau_dp = 0.2, 
    k_range = (0.001, 0.01), k_mean = 0.001, k_dp = 0.005, lam_c = 4727.833, 
    lam_c_tol = 1000, labren_id = 1, by = "month", n_cycles = 3, start_at = 0, 
    repetitions = 10 ** 2, plot = True
    ):
    turtles_0 = create_turtles(
        n, tau_range, tau_mean, tau_dp, k_range, k_mean, k_dp
        )
    
    if (by == "month"):
        labels = [
            "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", 
            "Oct", "Nov", "Dec"
            ]
        
        labren_data = labren(labren_id)["ts"]
        cycle = 12
    else:
        labels = ["Summer", "Autumn", "Winter", "Spring"]
        labren_data = labren(labren_id, by = "season")["ts"]
        cycle = 4
    
    labels = reorder(labels, start_at)
    labren_data = reorder(labren_data, start_at)
    
    if not n_cycles == 1:
        labels_0, labren_data_0 = tuple(labels), tuple(labren_data)
        [labels.extend(labels_0) for i in range(n_cycles - 1)]
        [labren_data.extend(labren_data_0) for i in range(n_cycles - 1)]
    
    if repetitions == 0:
        turtles = {"unentrain": turtles_0}
        
        for i in range(cycle * n_cycles):
            key = list(turtles)[-1]
            lam = labren_data[i]
            turtles_i = entrain_turtles(
                turtles[key], turtles_0, lam, lam_c, lam_c_tol
                )
            turtles[labels[i].lower()] = turtles_i
    else:
        turtles_n = {}
        
        for i in range(repetitions + 1):
            turtles_i = {"unentrain": turtles_0}
            
            for j in range(cycle * n_cycles):
                key = list(turtles_i)[-1]
                lam = labren_data[j]
                turtles_j = entrain_turtles(
                    turtles_i[key], turtles_0, lam, lam_c, lam_c_tol
                    )
                turtles_i[labels[j].lower()] = turtles_j
            
            turtles_n["r_" + str(i + 1)] = turtles_i
        
        turtles = average_turtles(turtles_n)

    if plot == True: plot_model_density(
        turtles, lam_c, labren_id, n_cycles, repetitions
        )
    
    return(turtles)

def plot_model_density(turtles, lam_c, labren_id, n_cycles, repetitions):
    if len(turtles) == 13:
        colors = sns.color_palette("tab10", 12)
    else:
        colors = ["#f98e09", "#bc3754", "#57106e", "#5ec962"]
    
    n = len(turtles[list(turtles)[0]])
    lat = labren(labren_id)["lat"]
    start = list(turtles)[1].title()
    labels = [i.title() for i in list(turtles)]
    
    title = ("N = ${n}$, $\\lambda_c = {lam_c}$, Latitude = ${lat}$, " +\
             "Cycles = ${n_cycles}$, Start = {start}, " +\
             "Repetitions = ${repetitions}$")\
             .format(n = str(n), lam_c = str(lam_c), lat = str(lat), 
              n_cycles = str(n_cycles), start = start, 
              repetitions = str(repetitions))
    
    plt.rcParams.update({'font.size': 10})
    plt.clf()
    
    fig, ax = plt.subplots()
    
    for i, j in enumerate(turtles):
        tau_i = [i["tau"] for i in np.array(turtles[j])]
        n = len(tau_i)
        
        if (i == 0):
            color = "black"
            linewidth = 3
        else:
            color = colors[i - 1]
            linewidth = 1
        
        sns.kdeplot(tau_i, color = color, label = labels[i], 
                    linewidth = linewidth, warn_singular = False)

    ax.set_xlabel("$\\tau$")
    ax.set_ylabel("Kernel Density Estimate (KDE)")
    ax.set_xlim(23.5, 24.6)
    ax.set_title(title, fontsize = 8)
    
    plt.legend(fontsize = 8)
    plt.show()

def average_turtles(turtles_n):
    n = len(turtles_n)
    keys = list(turtles_n[list(turtles_n)[0]])
    out = {}
    
    for i in keys: # i = 0 => "unentrain"
        turtles_i = []
        
        for a, b in enumerate(turtles_n): # a = 0 => b = "r_1"
            turtles_i.append(turtles_n[b][i])
        
        tau_i, k_i = [], []
        
        for a in range(len(turtles_i)):
            tau_i.append([b["tau"] for b in np.array(turtles_i[a])])
            k_i.append([b["k"] for b in np.array(turtles_i[a])])
        
        tau_i = functools.reduce(lambda x, y: np.array(x) + np.array(y), 
                                 tau_i) / n
        k_i = functools.reduce(lambda x, y: np.array(x) + np.array(y), 
                               k_i) / n
        
        turtles_i = []
        
        for a in range(len(tau_i)):
            turtles_i.append({"tau": tau_i[a], "k": k_i[a]})
        
        out[i] = turtles_i
        
    return(out)

def reorder(x, start_at = 0):
    n = len(x)
    
    if (not start_at == 0):
        order = list(np.arange(start_at, n))
        order.extend(list(np.arange(0, start_at)))
        
        return([x[i] for i in order])
    else:
        return(x)

# labren(72272)
x = run_model(
    labren_id = 72272, by = "season", n_cycles = 2, repetitions = 100
    )

# labren(1)
x = run_model(labren_id = 1, by = "season", n_cycles = 2, repetitions = 100)
